### ISSUE TITLE
Bulk requests

### ISSUE THREAD

**Issue by burgerni10** [OptimistikSAS/OIBus#1079]
Add a new (tab) menu to OIBus for bulk requests

On this page, the user must choose a protocol that can handle historians (see issue #1078).

Once the protocol selected, its UI is displayed for the user (minus the scan mode because here the request is done once).

The user can select the startDate and endDate of the request, fill the points (OPC-HDA / OPCUA) or the SQL request (SQLDbToFile).

A run button must be added to execute the bulk request, and this request must be executed only once by the engine.
This request may be split to manage large amount of data (time split and / or point split) and the execution of the request chunks can be monitor through logs

**Comment by kukukk** [comment]
What happens if OIBus is restarted during the bulk action?
Do we need to keep track of the progress and store it in config/database so we can continue on the next start?
Do we allow setting up multiple bulk request? Or the run button should be disabled until the bulk request is running?
What if there is an overlap in time with the normal mode? I mean, lastCompletedAt for normal mode is 1 February 2021 (there was a downtime period on the server and OIBus was not running) and endDate for bulk is set to 4 February 2021?

**Comment by burgerni10** [comment]
_What happens if OIBus is restarted during the bulk action?_
->We keep track of the bulk action. For example, we request one year. The request is split in 52 sub request (a week length each). If OIBus is restarted after sub-request 20 is completed, it can continue at sub-request 21. This may allow to track the progress on screen (progress bar for example ?)

_Do we allow setting up multiple bulk request? Or the run button should be disabled until the bulk request is running?_
->I think we should disable the button when another bulk request is running to avoid performance issue. When anotherb ulk request is running, the button is disable and a link to the current bulk request (and/or its completion percentage) can be displayed under the button

_What if there is an overlap in time with the normal mode? I mean, lastCompletedAt for normal mode is 1 February 2021 (there was a downtime period on the server and OIBus was not running) and endDate for bulk is set to 4 February 2021?_
->The bulk request results may be used differently (for backup for example). I don't see that as a problem.

**Comment by kukukk** [comment]
What do we do with the collected data?
Handle the same way as during normal operation and send to every North capable of handling it (this is the case when when normal mode and bulk mode can overlap)?
Or store in a specific format (CSV?) and let the user decide what to do with it?

**Comment by burgerni10** [comment]
Fore now, store the data in a folder specified by the user.

If we want to send the data somewhere, a folder scanner may be used after. 

Later, we may improve the bulk request by asking the user how and where to send the data directly from the bulk request interface. But doing that now may be too much at once.

**Comment by jfhenon** [comment]
Quick comment. 

We should use the exact same principle as HDA bulk. We save to local tmp files, rename them to CSV when it's finished, a restart will check existence of CSV files before requesting the history etc... 

When this is done, the user can activate a process which will send file per file to a North handler then move them or delete them like on success (maybe reusing for the south scan folder)

**Comment by burgerni10** [comment]
As for South/North handler, the bulk page list the bulk requests and offers the user to create a new one.

The same UI can be used for each history-supported protocols, without scan mode / scan groups but with start / end dates instead.

It may be useful to keep track of the datetime where the request is run and to display it on the UI (last run : none or last run : Tuesday, February 23 2021 at 14:54:30).

**Comment by jfhenon** [comment]
> Add a new (tab) menu to OIBus for bulk requests
> 
> On this page, the user must choose a protocol that can handle historians (see issue #1078).
> 
> Once the protocol selected, its UI is displayed for the user (minus the scan mode because here the request is done once).
> 
> The user can select the startDate and endDate of the request, fill the points (OPC-HDA / OPCUA) or the SQL request (SQLDbToFile).
> 
> A run button must be added to execute the bulk request, and this request must be executed only once by the engine.
> This request may be split to manage large amount of data (time split and / or point split) and the execution of the request chunks can be monitor through logs

I think it may be easier to select one of the South instance (that supports historical data) rather than a protocol.
The advantage would be that the South already contains all informations necessary to connect to the source so we can reuse it. This would mean we don't need to re-enter the configuration (server, user, pass, etc...).

the next question is: do we want to reuse the list of points/query of the South protocol? My answer is yes. Not only, it's very easy again for the user and, if necessary, it is always possible to modify this list for the bulk and restore it later with export/import. It's also possible to create a new South instance (just for the sake of launching a bulk mode). 

So basically, the screen would only have to select the list of Instance (supporting historical queries), start/end  dates and other required parameters to throttle the bulk mode.

**Comment by jfhenon** [comment]
One of the issue is that there is significant difference of behavior between SQL and other (points) protocols. I wait to see the design but we may have to differentiate the approaches. For example, the HDA/HA protocols can work one point at a time like for HDA. I don't think we can do this for SQL.
